module/zstd/include/string.h

```diff
@@ -55,6 +55,27 @@ extern "C" {
 #include_next <string.h>
 #endif /* _KERNEL */
 
+/*
+ * GCC/Clang doesn't know that the kernel's memcpy (etc)
+ * are standards-compliant so it can't be sure that it
+ * can inline its own optimized copies without altering
+ * behavior.  This is particularly significant for
+ * zstd which, for example, expects a fast inline memcpy
+ * in its inner loops.  Explicitly using the __builtin
+ * versions permits the compiler to inline where it
+ * considers this profitable, falling back to the
+ * kernel's memcpy (etc) otherwise.
+ */
+#if defined(_KERNEL) && defined(__GNUC__) && __GNUC__ >= 4
+#define	ZSTD_memcpy(d, s, l) __builtin_memcpy((d), (s), (l))
+#define	ZSTD_memmove(d, s, l) __builtin_memmove((d), (s), (l))
+#define	ZSTD_memset(p, v, l) __builtin_memset((p), (v), (l))
+#else
+#define	ZSTD_memcpy(d, s, l) memcpy((d), (s), (l))
+#define	ZSTD_memmove(d, s, l) memmove((d), (s), (l))
+#define	ZSTD_memset(p, v, l) memset((p), (v), (l))
+#endif
+
```

Comment:
I have to ask what was done here to confirm compiler builtins are not used as is. Note general kernel code in both FreeBSD and Linux *is* using them, so something special would have to be happening in zstd code for that to not happen.

I did a quick check by adding this on FreeBSD:
`
diff --git a/sys/contrib/openzfs/module/zstd/lib/common/xxhash.c b/sys/contrib/openzfs/module/zstd/lib/common/xxhash.c
index 597de18fc895..c3254a1ce646 100644
--- a/sys/contrib/openzfs/module/zstd/lib/common/xxhash.c
+++ b/sys/contrib/openzfs/module/zstd/lib/common/xxhash.c
@@ -86,6 +86,9 @@ static void  XXH_free  (void* p)  { free(p); }
 #include <string.h>
 static void* XXH_memcpy(void* dest, const void* src, size_t size) { return memcpy(dest,src,size); }
 
+void foomemset(char *buf);
+void foomemset(char *buf) { memset(buf, 0, 20); }
+
 #ifndef XXH_STATIC_LINKING_ONLY
 #  define XXH_STATIC_LINKING_ONLY
 #13904 
`
... and confirmed foomemset is using the builtin at least on that platform, I don't have handy manner to check Linux right now.

Regardless of what's going on, replacing memcpy et al uses with ZSTD_ prefixed variants is in my opinion too problematic -- for one one will have to keep a look out on each update.

In absolute worst case this can redefine these routines, but as noted previously this should not even be necessary. If things fail on Linux, there is probably a missing include.

Reply:
> I have to ask what was done here to confirm compiler builtins are not used as is.
The kernel's out-of-line memcpy() etc were hot on perftop; their being visible at all there means they're non-builtins.
> replacing memcpy et al uses with ZSTD_ prefixed variants is in my opinion too problematic -- for one one will have to keep a look out on each update.
Upstream zstd already did this